lib/galaxy/metadata/__init__.py

```diff
@@ -0,0 +1,277 @@
+        """Load metadata calculated externally into specified dataset."""
+
+
+class JobExternalOutputMetadataWrapper(MetadataCollectionStrategy):
+    """
+    Class with methods allowing set_meta() to be called externally to the
+    Galaxy head.
+    This class allows access to external metadata filenames for all outputs
+    associated with a job.
+    We will use JSON as the medium of exchange of information, except for the
+    DatasetInstance object which will use pickle (in the future this could be
+    JSONified as well)
+    """
+
+    def __init__(self, job_id):
+        self.job_id = job_id
+
+    def _get_output_filenames_by_dataset(self, dataset, sa_session):
+        if isinstance(dataset, galaxy.model.HistoryDatasetAssociation):
+            return sa_session.query(galaxy.model.JobExternalOutputMetadata) \
+                             .filter_by(job_id=self.job_id,
+                                        history_dataset_association_id=dataset.id,
+                                        is_valid=True) \
+                             .first()  # there should only be one or None
+        elif isinstance(dataset, galaxy.model.LibraryDatasetDatasetAssociation):
+            return sa_session.query(galaxy.model.JobExternalOutputMetadata) \
+                             .filter_by(job_id=self.job_id,
+                                        library_dataset_dataset_association_id=dataset.id,
+                                        is_valid=True) \
+                             .first()  # there should only be one or None
+        return None
+
+    def _get_dataset_metadata_key(self, dataset):
+        # Set meta can be called on library items and history items,
+        # need to make different keys for them, since ids can overlap
+        return "%s_%d" % (dataset.__class__.__name__, dataset.id)
+
+    def invalidate_external_metadata(self, datasets, sa_session):
+        for dataset in datasets:
+            jeom = self._get_output_filenames_by_dataset(dataset, sa_session)
+            # shouldn't be more than one valid, but you never know
+            while jeom:
+                jeom.is_valid = False
+                sa_session.add(jeom)
+                sa_session.flush()
+                jeom = self._get_output_filenames_by_dataset(dataset, sa_session)
+
+    def setup_external_metadata(self, datasets_dict, sa_session, exec_dir=None,
+                                tmp_dir=None, dataset_files_path=None,
+                                output_fnames=None, config_root=None,
+                                config_file=None, datatypes_config=None,
+                                job_metadata=None, compute_tmp_dir=None,
+                                include_command=True, max_metadata_value_size=0,
+                                kwds=None):
+        kwds = kwds or {}
+        if tmp_dir is None:
+            tmp_dir = MetadataTempFile.tmp_dir
```

Comment:
This (and below) looks so weird and unnecessary to me. I know it's old code, but Is it worth putting at least a `# FIXME` in there ?

Reply:
Yeah... this seems bad... if it is used at all it is almost worse - it would seem conflicting jobs might overwrite each other here...

Reply:
Going to let the tests run on https://github.com/galaxyproject/galaxy/pull/7213 and see what happens, I'm not sure we have any tests that would indicate if there is a problem here though.